RFC-0051: EXCLUDE Clause (#51)
partiql/partiql-docs: RFCs/0051-exclude-operator.adoc

Reviewer: Top-level, let's format these lines to be like 80 or 120 characters wide.

= Exclude Clause

:markup-in-source: verbatim,quotes

* Start Date: 2023-11-07
* PartiQL Issue: https://github.com/partiql/partiql-spec/issues/39
* RFC PR: https://github.com/partiql/partiql-docs/pull/51

== Summary

This doc defines the `EXCLUDE` binding tuple operator used to omit nested values before projection and defines the semantics in terms of syntactic rewrites with existing operators.

== Motivation

> SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[slack overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.

Author: Ah whoops. I'll fix.

```suggestion
SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[Stack Overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.
```

----
-- For the value `r` in our `SELECT VALUE` struct:
-- Assuming `<v~n-1~>` is the identifier created from the previous exclude step, `s~n-1~`
SELECT VALUE {
    'r':
        CASE
            WHEN ... --  branch(es) dependent on ``s~1~``'s rewrite rule
                    ... -- nested `CASE` expressions for `s~2~...s~n~`
                        CASE
                            WHEN ... -- branch(es) dependent on ``s~n~``'s rewrite rule
                        ELSE <v~n-1~>
                        END
            ELSE r
        END
}
----

[[anchor-2]]
.Rewrite rule 2: `EXCLUDE` steps `s~1~,...,s~n-1~`
For this rewrite rule definition, let `<v~i-1~>` be the identifier created from the previous exclude step (or `r` if this is the first step). For some exclude step `s~i~` that is not the last step, we case on the type of exclude step.

[[anchor-2ai]] Rule 2.a.i::
    If `s~i~` is a case-sensitive tuple attribute exclude step (e.g. `."foo"` or `['foo']`), where `<v~i~>` and `<attr~i~>` are fresh variables, add the following `WHEN` branch to the `i`^th^ nested `CASE`.
[source,partiql,subs="+{markup-in-source}"]
----
WHEN <v~i-1~> IS STRUCT THEN

>     PIVOT (

Author: Brought up in a separate conversation w/ Josh, may need to surround the `PIVOT ... FROM UNPIVOT ... ` (or any other subquery) in parens for SQL compatibility. I'll add parens to these subqueries.

Author: Resolved in: https://github.com/partiql/partiql-docs/pull/51/commits/3910333bf0c850efafc177e205cecd94c222f764

        CASE 
            WHEN <attr~i~> = <s~i~> THEN
                -- Apply rewrite rules on remaining exclude steps `s~i+1~,...,s~n~`
            ELSE <v~i~>
        END
    )
    AT <attr~i~>
    FROM UNPIVOT <v~i-1~> AS <v~i~> AT <attr~i~>
----
[[anchor-2aii]] Rule 2.a.ii::
    If `s~i~` is a case-insensitive tuple attribute exclude step (e.g. `.foo`), where `<v~i~>` and `<attr~i~>` are fresh variables, add the following `WHEN` branch to the the `i`^th^ nested `CASE`.
[source,partiql,subs="+{markup-in-source}"]
----

> WHEN <v~i-1~> IS STRUCT THEN

Author: (similar to other comment regarding `IS ARRAY`), I'll fix in next revision to use `IS TUPLE`.

2. paths of length greater than `i` (i.e. have additional steps)

If there are any `EXCLUDE` paths of length `i`, then similar to <<anchor-3ai, Rule 3.a.i>> and <<anchor-3aii, Rule 3.a.ii>>, we add a `WHERE` clause to filter out those fields. The fields to exclude will be grouped together based on if the tuple attribute exclude step was case sensitive or case-insensitive.

If there are any `EXCLUDE` paths of length greater than `i`, then similar to <<anchor-2ai, Rule 2.a.i>> and <<anchor-2aii, Rule 2.a.ii>>, we add a `CASE` expression within the `PIVOT`. This `CASE` expression within the `PIVOT` will define a `WHEN` branch for each of the unique tuple attribute steps. Each of these `WHEN` branches will apply the rewrite rules for the exclude paths that have additional steps and equivalent tuple attribute or tuple wildcard. An `ELSE` branch will be added to this `CASE` expression which will apply the rewrite rules for the exclude paths with a tuple wildcard at level `i` and additional steps.
[source,partiql,subs="+{markup-in-source}"]
----
-- Let `k` represent the number of unique exclude tuple attrs for paths of length
-- greater than `i`.

> -- `<v~i~>` and `<attr~i~>` are fresh variables

Author: I'll take another look at the variables in these multi-exclude path rewrites. Agree that it can be confusing w/ how it's currently written.

    ]
  },
  {
    'a': {
      'foo': 6,
      'baz': 8
    }
  },
  {
    'a': {
      'a1': {
        'foo': 9,
        'baz': 11
      },
      'a2': {
        'foo': 12,
        'baz': 14
      }
    }
  }
>>
----

== Drawbacks

`EXCLUDE` (or similar clause) is not part of the SQL or SQL++ standard. If `EXCLUDE` is added in a future standard, it's possible the syntax and semantics may change.

== Rationale and alternatives
[qanda]
In the original spec issue (https://github.com/partiql/partiql-spec/issues/39[partiql-spec#39]), `EXCEPT` was included as the keyword for this clause. Why was the keyword `EXCLUDE` chosen?::

`EXCLUDE` was chosen over `EXCEPT` since `EXCEPT` could be confused with the set/bag operator `EXCEPT`. `EXCLUDE` was also chosen by the SQL++ implementation, AsterixDB, through some similar reasoning:
+
[quote,https://issues.apache.org/jira/browse/ASTERIXDB-3059]
____
'EXCLUDE' (used in lieu of 'EXCEPT' to avoid confusion with the set operation)
____
+
Also of the databases sampled that have a similar clause (see <<Prior art>>), more had chosen `EXCLUDE` over `EXCEPT`.

Why is `EXCLUDE` modeled as a binding tuple operator as opposed to a value expression?::

> We had also considered modeling `EXCLUDE` as a value operation evaluated after the `<select clause>`. Evaluating `EXCLUDE` last could contradict the PartiQL specification's assertion that the `<select clause>` is evaluated last, which may add confusion. There were also some additional edge cases that complicated defining `EXCLUDE` as a value operator. For example, let's look at the following query:

Author: > Did you explore `EXCLUDE` as a value operation (maybe just a special-form) evaluated during the `<select clause>`? Similar to how any other special-form function invocation would work in the projection list?

Could you clarify what you mean by special-form function? Is this referring to non-standard function syntax (kinda like what substring, overlay do)?

Regarding modeling `EXCLUDE` as a value operation/function evaluated within the `SELECT` clause, we had also explored this option. These aren't quite functions though, since function arguments expect value expressions and exclude paths do not return values.

The above way you had modeled expressions evaluated left-to-right could be problematic. Consider `SELECT EXCLUDE t.a[0], EXCLUDE t.a[1].field FROM ...`. Should the second exclude path remove `field` from the `a`'s original `1` index or after evaluating the first exclude path (i.e. originally the `2` index). We found considering the paths as associative and evaluating the paths in parallel (i.e. applying all exclude paths on the same value/binding) as more intuitive.

> We had also considered modeling `EXCLUDE` as a value operation evaluated after the `<select clause>`. Evaluating `EXCLUDE` last could contradict the PartiQL specification's assertion that the `<select clause>` is evaluated last, which may add confusion. There were also some additional edge cases that complicated defining `EXCLUDE` as a value operator. For example, let's look at the following query:

Reviewer 2: 1. Yes, meaning non-standard function syntax.
2. Yes -- but that is the AST, which isn't the same as execution. We don't operate on the AST -- we convert `TRIM` to `$trim_leading` and `CURRENT_USER` to `$current_user`. The SQL:1999 Specification even talks about the conversion of the syntactic `BETWEEN` to comparison operators -- the theoretical `between` AST node disappears and gets replaced. We could just as easily say `<lhs> EXCLUDE ( <exclude-arg-1>, <exclude-arg-2> )` gets converted to `EXCLUDE(<lhs>, [ <exclude-arg-1-stringified>, <exclude-arg-2-stringified>])`. Anyways, it *could* provide flexibility, and I believe it's possible to model.
3. We could probably look at BigQuery which expects the RHS to be parenthesized. Ex: `SELECT * EXCEPT (a, b)`. I would imagine this query in BigQuery strips out both `a` and `b` simultaneously.

Given PartiQL's flexibility, I'm personally intrigued by the prospect of writing:
```
SELECT VALUE {
  'full_name': person.l_name || ', ' person.f_name,
  'encrypted_ssn': udf_to_encrypt(person.ssn),
  'all_other_details': person EXCLUDE (f_name, l_name, ssn)
} FROM Persons AS person
```

Especially since tuples are first-class citizens, I find it long overdue for some practical built-ins (beyond `TUPLEUNION` which PLK doesn't have implemented -- yet). It seems like a potential opportunity to allow for its introduction, especially since SQL's `SELECT` gets transformed to PartiQL's `SELECT VALUE <tuple>`.

> We had also considered modeling `EXCLUDE` as a value operation evaluated after the `<select clause>`. Evaluating `EXCLUDE` last could contradict the PartiQL specification's assertion that the `<select clause>` is evaluated last, which may add confusion. There were also some additional edge cases that complicated defining `EXCLUDE` as a value operator. For example, let's look at the following query:

Reviewer: Cannot be a value operator if you want to ever prune binding tuple variables.

= Exclude Clause

:markup-in-source: verbatim,quotes

* Start Date: 2023-11-07
* PartiQL Issue: https://github.com/partiql/partiql-lang/issues/27
* RFC PR: https://github.com/partiql/partiql-docs/pull/51

== Summary

> This doc defines the `EXCLUDE` binding tuple operator used to omit nested values before projection and defines the semantics in terms of syntactic rewrites with existing operators.

Reviewer 3: ```suggestion
This doc defines the `EXCLUDE` binding tuple operator used to omit (potentially) nested values before projection and defines the semantics in terms of syntactic rewrites into existing operators.
```

== Motivation

SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[Stack Overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.

> There is a similar need among PartiQL users to exclude certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.

Reviewer 4: ```suggestion
There is a similar need among PartiQL users to exclude certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields other than the field(s) that need to get excluded or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.
```

> There is a similar need among PartiQL users to exclude certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.

Reviewer 3: ```suggestion
There is a similar need among PartiQL users to exclude not only top-level, but also certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.
```

----
-- Suppose `tbl` is a collection of tuples that have `n` fields, `field~1~,...,field~n~`.
-- To filter out `field~i~`, we would have to list out all fields other than `field~i~`.
SELECT
    field~1~, ..., field~i-1~, field~i+1~, ..., field~n~ -- omit `field~i~` from tbl
FROM
    tbl
----

== Guide-level explanation

=== BNF Grammar

[source,ebnf]
----
<sfw query> ::=
    <select clause>
    <exclude clause>?  (* NEW *)
    <from clause>
    (* other clauses from <sfw query> in PartiQL spec figure 3 *)

<fws query> ::=
    <from clause>
    (* other clauses same as fws query *)
    <exclude clause>?  (* NEW *)
    <select clause>
    
<exclude clause> ::=
    EXCLUDE <exclude path> [, <exclude path>]...

<exclude path> ::=
    <identifier>
  | <exclude path> <dot> <identifier>
  | <exclude path> <left bracket> <int literal> <right bracket>
  | <exclude path> <left bracket> <string literal> <right bracket>
  | <exclude path> <dot> <asterisk>
  | <exclude path> <left bracket> <asterisk> <right bracket>

<identifier> ::= (* See identifier in PartiQL spec figure 4 *)

<asterisk> ::= "*"
<dot> :: = "."
<left bracket> ::= "["
<right bracket> ::= "]"
----

> NOTE: Despite their similar syntax and naming, ``<exclude path>``s are different than PartiQL path expressions.

Reviewer 4: Text can also include the spec section and link for Path expressions.

=== Terminology
* For an `<exclude path>`, we refer to the leftmost identifier as the 'root' and the other exclude path components as 'steps'.
+
[source]
----
e.g. tableFoo.a[1].*[*].b['c']
    |  root  |     steps     |
----
+
* We refer to the exclude steps as follows
** `.<identifier>` - tuple attribute exclude step
*** `.<quoted identifier>`/`[<string literal>]` - case-_sensitive_ tuple attribute
*** `.<unquoted identifier>` - case-_insensitive_ tuple attribute
** `[<int literal>]` - collection index exclude step
** `.*` - tuple wildcard exclude step
** `[*]` - collection wildcard exclude step

=== Out of scope / assumptions

* We restrict tuple attribute exclude steps to use string literals and collection index exclude steps to use int literals. Thus `<exclude paths>` are statically known. We can decide whether to add other exclude paths (e.g. expressions) if a use case arises.

> * If sufficient schema is present and the path can be resolved, we assume the root of an `EXCLUDE` path can be omitted. The variable resolution rules follow what is already included in the PartiQL specification.

Reviewer: I think we might want to have an example of attribute as a variable.

> * We require that every fully-qualified `<exclude path>` contain a root and at least one step. If a use case arises to exclude a binding tuple variable, then this functionality can be added.

Reviewer: What is the rationale for this limitation?  We should put that here.

* S-expressions are part of the Ion type system.footnote:[https://amazon-ion.github.io/ion-docs/docs/spec.html#sexp]

> PartiQL should support s-expression types and values since PartiQL's type system is a superset over the Ion types. Because the current PartiQL specification does not formally define s-expressions operations, we consider the definition of collection index and wildcard steps on s-expressions as out-of-scope for this RFC.

Reviewer 4: Perhaps the statement can be less assertive; I know this is one of those hotly debated topics. The spec. says:
> PartiQL’s data model extends SQL to Ion’s type system to cover schema-less and nested data. Such values can be
directly quoted with `quotes.

So text can just convey the message that s-expressions semantics as a collection type is not fully defined yet, hence is out of the scope.

Reviewer 3: Agreed. This statement makes more assertions about the PartiQL value system than does the spec.

> PartiQL should support s-expression types and values since PartiQL's type system is a superset over the Ion types. Because the current PartiQL specification does not formally define s-expressions operations, we consider the definition of collection index and wildcard steps on s-expressions as out-of-scope for this RFC.

Reviewer 3: I would reword this section. I would avoid usage of terms like 'we'  and write it more formally and more tersely.

e.g.,

```
=== Limitations

* This RFC requires that every fully-qualified `<exclude path>` contain a root and at least one step.
* This RFC restricts tuple attribute exclude steps to use string literals and collection index exclude steps to use int literals. Thus `<exclude paths>` are statically known. This 
* This RFC makes no changes to schema and name inference, and assumes that such inference is run as a prerequisite.
* This RFC defines `<exclude paths>` only over `list`, `bag`, and `tuple` value collections.

```

=== Rewrite Procedure
==== Step 1: subsumption of `EXCLUDE` paths
We perform the following step to ensure that there are no redundant `EXCLUDE` paths. That is, there is no path such that all of its excluded binding tuple values are excluded by another exclude path.footnote:[This subsumption step is included to make the subsequent rewrite steps easier to reason about. In a query without redundant exclude paths, this step is not necessary.]

For each `<exclude path>` `p=root~p~s~1~...s~x~`, we compare it with all other ``<exclude path>``s. `<exclude path>` `p` is said to be subsumed by another path `q=root~q~t~1~...t~y~` and not included in the rewritten `EXCLUDE` clause if any of the following rules apply:

NOTE: The following rules assume `root~p~=root~q~`.

> .Subsumption rules

Reviewer 4: I know we have the `table 1.` for the examples, but adding an example for each of the rules would also enhance readability.

[[anchor-1a]] Rule 1.a::
    If `y = 0` (i.e. `q` has no steps), `q` subsumes `p`.
[[anchor-1b]] Rule 1.b::
    If `x ≥ y` and `s~1~...s~x~=t~1~...t~x~`, `q` subsumes `p`. Put another way if `p` has at least as many steps as `q` and the steps up to ``q``'s length are equivalent, `q` subsumes `p`.

Otherwise, there must be some step at which `p` and `q` diverge. Let's call this step's index `i`.

[[anchor-1c]] Rule 1.c::

>     If `s~i~` is a tuple attribute and `t~i~` is a tuple wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.

Reviewer 4: For this and others that apply: In order for the underlying subsumption to happen (I.e., `t~i+1~...t~y~` subsumes `s~i+1~...s~x~`) the two should be considered as independent `<exclude path>`s so that they can have root; is that so? if yes, text may need to clarify that further.

>     If `s~i~` is a tuple attribute and `t~i~` is a tuple wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.

Reviewer 4: ```suggestion
    Given `i < y and i < x`, if `s~i~` is a tuple attribute and `t~i~` is a tuple wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

[[anchor-1d]] Rule 1.d::

>     If `s~i~` is a collection index and `t~i~` is a collection wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.

Reviewer 4: ```suggestion
    Given `i < y and i < x`, if `s~i~` is a collection index and `t~i~` is a collection wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

[[anchor-1e]] Rule 1.e::

>     If `s~i~` is a case-sensitive tuple attribute and `t~i~` is a case-insensitive tuple attribute and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.

Reviewer 4: ```suggestion
    Given `i < y and i < x`, if `s~i~` is a case-sensitive tuple attribute and `t~i~` is a case-insensitive tuple attribute and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

.Subsumption Examples
[options="header,footer"]
|=======================
|Exclude Path `p`|Exclude Path `q`|Notes
|`s.a`        |`t.a`       |No subsumption rules apply (roots differ)
|`t.a`        |`t.b`       |No subsumption rules apply
|`t.a.b.c`    |`t.a.*.d`   |No subsumption rules apply
|`t.a`        |`t`         |`q` subsumes `p` (by <<anchor-1a, 1.a>>)
|`t.a.b.c`    |`t.a.b.c`   |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
|`t.a.b.c`    |`t.a.b`     |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
|`t.a.b.c`    |`t.a.b.*`   |`q` subsumes `p` (by <<anchor-1c, 1.c>> then <<anchor-1a, 1.a>>)
|`t.a.b.c`    |`t.a.*.c`   |`q` subsumes `p` (by <<anchor-1c, 1.c>> then <<anchor-1b, 1.b>>)
|`t.a.b[1]`   |`t.a.b`     |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
|`t.a.b[1]`   |`t.a.b[*]`  |`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1a, 1.a>>)
|`t.a.b[1].c` |`t.a.b[1]`  |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
|`t.a.b[1].c` |`t.a.b[*]`  |`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1a, 1.a>>)
|`t.a.b[1].c` |`t.a.b[*].c`|`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1b, 1.b>>)
|`t.a."b"`    |`t.a.b`     |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1a, 1.a>>)

> |`t.a."b".c`  |`t.a.b.c`   |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1b, 1.b>>)

Reviewer 4: Would be good to have a 'No subsumption rule apply` for case sensitive mismatch.

|=======================

---
We first illustrate the rewrite rule for a single `EXCLUDE` path and then explain the syntax rewrite for multiple exclude paths.

> ==== Step 2 (single): rewrite a single `EXCLUDE` path

Reviewer 4: Considering that there are many rules, perhaps a pseudo-code accompanying the text explanation could also be helpful.

If the other clauses include an `ORDER BY`, we convert the top-level query back into an array by adding a position variable (i.e. `AT` clause) along with an `ORDER BY` over the position variable.

[source,partiql,subs="+{markup-in-source}"]
----
-- Original query:
<select clause>
EXCLUDE r.s~1~...s~n~
<from clause>
<other clauses>

-- Rewritten to:
<select clause>
FROM (
    SELECT VALUE {

>         'r': -- Apply below rewrite rules for steps `s~1~...s~n~`

Reviewer 4: ```suggestion
        'r': -- Apply rewrite rules explained in the following sections for steps `s~1~...s~n~`
```

>         ...  -- Other vars created from the other clauses

Reviewer 4: ```suggestion
        ...  -- Add other variables created from the other clauses using identity function
```

    }
    <from clause>
    <other clauses>
)
[   -- Include conversion back to array if `ORDER BY` present in `<other clauses>`
    -- Assume `<topLevelTbl>` and `<idx>` are fresh variables
    AS <topLevelTbl> AT <idx>
    ORDER BY <idx>
]
----

> The main idea for rewriting the `EXCLUDE` steps `s~1~,...,s~n~` is to create a nested `CASE` expression for each step, whereby the nested `CASE` expressions for `s~1~,...,s~n-1~` unnest the input binding tuple and the final `CASE` expression for `s~n~` (i.e. the final step) filters out the desired tuple field(s) or collection index(es). Every exclude step has an expected type to process during evaluation. Tuple attribute and wildcard exclude steps expect a tuple. Whereas a collection index expects an array and a collection wildcard expects an array or bag. The `CASE` expression at each level `i` recreates this expected type by including a `WHEN` branch based on the expected type. Each `CASE` expression will include an `ELSE` branch which outputs the previous level's identifier. This set of branches ensures that at evaluation time, if there is a type mismatch (e.g. evaluation value is an array while the exclude step is a tuple attribute), there is no evaluation error and the previous level's value is returned through the `ELSE` branch. This behavior applies to both the permissive and strict typing modes.

Reviewer 4: ```suggestion
The main idea for rewriting the `EXCLUDE` steps `s~1~,...,s~n~` is to create a nested `CASE` expression for each step, whereby the nested `CASE` expressions for `s~1~,...,s~n-1~` unnest the input binding tuple and the final `CASE` expression for `s~n~` (i.e. the final step) filters out the desired tuple field(s) or collection index(es). Every exclude step has an expected type to process during evaluation. Tuple attribute and wildcard exclude steps expect a `Tuple`, whereas a collection index expects `Array` and a collection wildcard expects `Array` or `Bag` types. The `CASE` expression at each level `i` recreates this expected type by including a `WHEN` branch based on the expected type. Each `CASE` expression will include an `ELSE` branch which outputs the previous level's identifier. This set of branches ensures that at evaluation time, if there is a type mismatch (e.g. evaluation value is an array while the exclude step is a tuple attribute), there is no evaluation error and the previous level's value is returned through the `ELSE` branch. This behavior applies to both the permissive and strict typing modes.
```

Some examples:
[source,sql]
----
-- select all referencable columns from all tables except t2.c4
> SELECT * EXCEPT(c4) FROM VALUES(1, 2) AS t1(c1, c2), VALUES(3, 4) AS t2(c3, c4);
  1   2   3

-- select all referencable columns from a table, except a nested field.
> SELECT * EXCEPT(c2.b) FROM VALUES(1, named_struct('a', 2, 'b', 3)) AS t(c1, c2);
  1  { "a" : 2 }

-- Removing all fields results in an empty struct
> SELECT * EXCEPT(c2.b, c2.a) FROM VALUES(1, named_struct('a', 2, 'b', 3)) AS t(c1, c2);
  1  { }

-- Overlapping names result in an error
> SELECT * EXCEPT(c2, c2.a) FROM VALUES(1, named_struct('a', 2, 'b', 3)) AS t(c1, c2);
  Error: EXCEPT_OVERLAPPING_COLUMNS
----

=== MongoDB
* Reference: https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/#return-all-but-the-excluded-fields
* More info on MongoDB: https://dbdb.io/db/mongodb

MongoDB supports excluding certain fields by setting the attribute's value to `0`. The following returns all fields other than the `status` and `instock` field for all of the matching documents.

[source,shell]
----
db.inventory.find( { status: "A" }, { status: 0, instock: 0 } )
----

> == Unresolved questions

Reviewer 4: Wondering if partial schema case be out of scope only? i.e. having SQL's schema-full within the scope as well.
